// sanity-plugin-seo-pane/lib/Feedback.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Feedback;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ui = require("@sanity/ui");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Feedback(_ref) {
  var children = _ref.children,
      isError = _ref.isError,
      padding = _ref.padding;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: padding
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 3,
    radius: 2,
    shadow: 1,
    tone: isError ? "critical" : "caution"
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 2
  }, children)));
}

Feedback.propTypes = {
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node]).isRequired,
  isError: _propTypes.default.bool,
  padding: _propTypes.default.number
};
Feedback.defaultProps = {
  isError: false,
  padding: 3
};
