// sanity-plugin-seo-pane/lib/SeoPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SeoPane;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactQuery = require("react-query");

var _SeoPaneComponent = _interopRequireDefault(require("./SeoPaneComponent"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var queryClient = new _reactQuery.QueryClient();

function SeoPane(_ref) {
  var displayed = _ref.document.displayed,
      options = _ref.options;
  return /*#__PURE__*/_react.default.createElement(_ui.ThemeProvider, {
    theme: _ui.studioTheme
  }, /*#__PURE__*/_react.default.createElement(_reactQuery.QueryClientProvider, {
    client: queryClient
  }, /*#__PURE__*/_react.default.createElement(_SeoPaneComponent.default, {
    document: displayed,
    options: options
  })));
}

SeoPane.propTypes = {
  document: _propTypes.default.shape({
    displayed: _propTypes.default.shape({
      _id: _propTypes.default.string,
      _rev: _propTypes.default.string
    })
  }).isRequired,
  options: _propTypes.default.shape({
    keywords: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]).isRequired,
    synonyms: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]),
    url: _propTypes.default.func.isRequired
  }).isRequired
};
