// sanity-plugin-seo-pane/lib/SerpPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SerpPreview;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ui = require("@sanity/ui");

var _Feedback = _interopRequireDefault(require("./Feedback"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MAX_CHARACTERS = 145;

function splitUrl(url) {
  var urlObj = new URL(url);
  var urlParts = [urlObj.origin];

  if (urlObj.pathname) {
    urlParts.push(...urlObj.pathname.split("/").filter(Boolean));
  } // Remove last part if it's long-ish
  // This is my guess as to what Google does


  if (urlParts[urlParts.length - 1].length > 30) {
    urlParts.pop();
  }

  return urlParts;
}

function SerpPreview(_ref) {
  var title = _ref.title,
      metaDescription = _ref.metaDescription,
      url = _ref.url;
  return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3,
    style: {
      maxWidth: 600
    }
  }, url ? /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, splitUrl(url).map((part, index) => /*#__PURE__*/_react.default.createElement(_ui.Text, {
    key: part,
    muted: index !== 0
  }, index === 0 ? part : "\u203A ".concat(part)))) : /*#__PURE__*/_react.default.createElement(_Feedback.default, {
    padding: 0
  }, "URL not found, your frontend may be missing a ", /*#__PURE__*/_react.default.createElement("code", null, "canonical"), " tag"), title ? /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 3,
    style: {
      color: '#1a0dab'
    }
  }, title) : null, metaDescription ? /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, metaDescription.length > MAX_CHARACTERS ? metaDescription.slice(0, MAX_CHARACTERS) + '...' : metaDescription) : null);
}

SerpPreview.propTypes = {
  metaDescription: _propTypes.default.string.isRequired,
  title: _propTypes.default.string.isRequired,
  url: _propTypes.default.string.isRequired
};
